skfp/similarities/tanimoto.py

```diff
@@ -0,0 +1,350 @@
+    },
+    prefer_skip_nested_validation=True,
+)
+def tanimoto_count_distance(
+    vec_a: Union[np.ndarray, csr_array], vec_b: Union[np.ndarray, csr_array]
+) -> float:
+    """
+    Computes the Tanimoto distance for binary data between two input arrays or sparse matrices
+    by subtracting similarity value from 1.
+    Calculated distance falls within the range from 0 to 1.
+    Passing all-zero vectors to this function results in distance of 0.
+
+    Parameters
+    ----------
+    vec_a : {ndarray, sparse matrix}
+        First binary input array or sparse matrix.
+
+    vec_b : {ndarray, sparse matrix}
+        Second binary input array or sparse matrix.
+
+    Returns
+    ----------
+    distance : float
+        Tanimoto distance between A and vec_b.
+
+    Examples
+    ----------
+    >>> from skfp.similarities import tanimoto_count_distance
+    >>> vec_a = np.array([7, 1, 1])
+    >>> vec_b = np.array([7, 1, 2])
+    >>> dist = tanimoto_count_distance(vec_a, vec_b)
+    >>> dist
+    0.02
+
+    >>> vec_a = csr_array(([7, 1, 1]))
+    >>> vec_b = csr_array(([7, 1, 2]))
+    >>> dist = tanimoto_count_distance(vec_a, vec_b)
+    >>> dist
+    0.02
+    """
+
+    return 1 - tanimoto_count_similarity(vec_a, vec_b)
+
+
+def _tanimoto_binary_numpy(vec_a: np.ndarray, vec_b: np.ndarray) -> float:
+    """
+    Calculates the Tanimoto similarity between two binary numpy arrays.
+    """
+    _check_nan(vec_a)
+    _check_nan(vec_b)
+
+    if vec_a.sum() == 0 == vec_b.sum():
+        return 1.0
+
+    dot_ab: float = np.dot(vec_a, vec_b)
+    dot_aa: float = np.sum(vec_a)
+    dot_bb: float = np.sum(vec_b)
+
+    denominator: float = dot_aa + dot_bb - dot_ab
+    _check_zero_denominator(denominator)
+
+    tanimoto_similarity = dot_ab / denominator
+    return tanimoto_similarity
+
+
+def _tanimoto_binary_scipy(vec_a: csr_array, vec_b: csr_array) -> float:
+    """
+    Calculates the Tanimoto similarity between two binary scipy arrays.
+    """
+    _check_nan(vec_a)
+    _check_nan(vec_b)
+
+    if np.sum(vec_a) == 0 and np.sum(vec_b) == 0:
+        return 1.0
+
+    a_and_b: float = vec_a.multiply(vec_b).sum()
+    a_sum: float = vec_a.sum()
+    b_sum: float = vec_b.sum()
+
+    denominator: float = a_sum + b_sum - a_and_b
+    _check_zero_denominator(denominator)
+
+    tanimoto_sim: float = a_and_b / denominator
+
+    return tanimoto_sim
+
+
+def _tanimoto_count_numpy(vec_a: np.ndarray, vec_b: np.ndarray) -> float:
+    """
+    Calculates the Tanimoto similarity between two count data numpy arrays.
+    """
+    _check_nan(vec_a)
+    _check_nan(vec_b)
+
+    if np.sum(vec_a) == 0 and np.sum(vec_b) == 0:
+        return 1.0
+
+    dot_ab: float = np.dot(vec_a, vec_b)
+    dot_aa: float = np.dot(vec_a, vec_a)
+    dot_bb: float = np.dot(vec_b, vec_b)
+
+    denominator: float = dot_aa + dot_bb - dot_ab
+    _check_zero_denominator(denominator)
+
+    tanimoto_sim: float = dot_ab / denominator
+
+    return tanimoto_sim
```

Comment:
It would be beneficial to consider [Numba JIT compilation](https://numba.readthedocs.io/en/stable/user/jit.html) to speed up such functions. Could you measure this, e.g. for nearest neighbors searches on HIV dataset and ECFP4 fingerprints?